azkaban-common/src/main/java/azkaban/executor/ExecutionFlowDao.java

```diff
@@ -255,17 +255,31 @@ private void updateExecutableFlow(final ExecutableFlow flow, final EncodingType
             + "SET status=?,update_time=?,start_time=?,end_time=?,enc_type=?,flow_data=? "
             + "WHERE exec_id=?";
 
-    final String json = JSONUtils.toJSON(flow.toObject());
     byte[] data = null;
     try {
+      // If this action fails, the execution must be failed.
+      final String json = JSONUtils.toJSON(flow.toObject());
       final byte[] stringData = json.getBytes("UTF-8");
       data = stringData;
       // Todo kunkun-tang: use a common method to transform stringData to data.
       if (encType == EncodingType.GZIP) {
         data = GZIPUtils.gzipBytes(stringData);
       }
     } catch (final IOException e) {
-      throw new ExecutorManagerException("Error encoding the execution flow.");
+      flow.setStatus(Status.FAILED);
+      updateExecutableFlowStatus(flow);
+      throw new ExecutorManagerException("Error encoding the execution flow. Execution Id  = "
+          + flow.getExecutionId());
+    } catch (final NullPointerException npe) {
+      flow.setStatus(Status.FAILED);
+      // Likely due to serialization error
+      if ( data == null) {
+        logger.warn("Failed to serialize executable flow for " + flow.getExecutionId()
+        + ". Likely due to SLA missing options. Please re-check the flow SLA");
+      }
+      updateExecutableFlowStatus(flow);
```

Comment:
This is invoked two times and can be moved to a finally block. Or do we want to update only during failure or exception?

Reply:
We only want to update during failure. The normal case is when we update multiple entries including flow_data